pycernan/avro/client.py

```diff
@@ -1,47 +1,133 @@
 """
     Base Avro client from which all other clients derive.
 """
+import contextlib
 import socket
 
 import pycernan.avro.config
 
 from abc import ABCMeta, abstractmethod
+from queue import Queue, Empty
 
-from pycernan.avro.exceptions import EmptyBatchException
+from pycernan.avro.exceptions import EmptyBatchException, EmptyPoolException
 from pycernan.avro.serde import serialize
 
 
+DefunctConnection = None
+
+
+class TCPConnectionPool(object):
+    """
+    Fork friendly TCP connection pool.
+
+    Adapted from: https://github.com/gevent/gevent/blob/master/examples/psycopg2_pool.py
+    """
+    def __init__(self, host, port, maxsize, connect_timeout, read_timeout):
+        if maxsize <= 0:
+            raise ValueError("maxsize must be > 0")
+
+        self.host = host
+        self.port = port
+        self.size = 0
+        self.maxsize = maxsize
+
+        self.connect_timeout = connect_timeout
+        self.read_timeout = read_timeout
+        self.pool = None
+
+    def create_connection(self):
+        sock = socket.create_connection((self.host, self.port), timeout=self.connect_timeout)
+        sock.settimeout(self.read_timeout)
+        return sock
+
+    def get(self, _block=True):
+        if not self.pool:
+            # Lazy queue creation, ensuring we are fork() friendly.
+            # Caveat - Users must not use the connection pool prior to
+            # forking.
+            self.pool = Queue()
+
+        if self.size >= self.maxsize or self.pool.qsize():
+            try:
+                existing_item = self.pool.get(_block)
+            except Empty:
+                # Expected to happen only when users override _block=False
+                raise EmptyPoolException()
+
+            if not existing_item:
+                try:
+                    existing_item = self.create_connection()
+                except Exception:
```

Comment:
I think here (and other places in this class) is a case where you really want a bare `except:` so as to avoid drift in the case of things like `KeyboardInterrupt` or horrors like string exceptions 😱